includes/class-wc-payments.php

```diff
@@ -686,7 +686,7 @@ public static function register_gateway( $gateways ) {
 
 			}
 
-			if ( is_add_payment_method_page() ) {
+			if ( is_add_payment_method_page() || ( self::$registered_card_gateway->is_subscriptions_enabled() && self::$registered_card_gateway->is_changing_payment_method_for_subscription() ) ) {
```

Comment:
Apologies, but I've just had another thought (sorry lol). 

I was wondering how we filtered the payment gateways when simply checking out for a subscription product (i.e. not on the add payment method page nor changing payment method for a subscription) and it's basically because we check whether the potential payment method `is_enabled_at_checkout()` over [here](https://github.com/Automattic/woocommerce-payments/blob/4bbebe49e4747a200d65fccded7bcc865f145e25/includes/payment-methods/class-upe-payment-gateway.php#L935). 

If we take a peek inside that function, we can see that [if there is currently a subscription item in the cart, we check whether the payment method is reusable](https://github.com/Automattic/woocommerce-payments/blob/4bbebe49e4747a200d65fccded7bcc865f145e25/includes/payment-methods/class-upe-payment-method.php#L113-L118) and otherwise the payment method is included. If we simply moved this additional logic, highlighted here inside the `register_gateway()` function, into the `UPE_Payment_Method::is_enabled_at_checkout()` function instead, then we wouldn't need to handle any additional logic with the `$reusable_methods` array in this function, since these payment methods are already filtered. 

FWIW, `is_subscriptions_enabled()` and `is_changing_payment_method_for_subscription()` are both available to the `UPE_Payment_Method` class, thanks to [the inclusion of the `WC_Payments_Subscriptions_Utilities` trait](https://github.com/Automattic/woocommerce-payments/blob/4bbebe49e4747a200d65fccded7bcc865f145e25/includes/payment-methods/class-upe-payment-method.php#LL24C6-L24C41), so I don't think there would be any issues with asking these questions within this class instead of relying on the gateway. 

Does that make sense? Does this sound like a good idea? Am I just a rotten trouble-maker? Please let me know, I'm all ears. 😁

Reply:
Thanks, I think it's a good point and worth moving to the `UPE_Payment_Method` - eventually the check ends up looking easier and in the right place. The main thing that concerned me with the previous solution was that `register_gateway` function seemed to be too high-level place for such filtering logic so I'm more than happy to change this.